### ISSUE TITLE
Using Swarm with Agents

### ISSUE THREAD

**Issue by pragdave** [bitwalker/swarm#62]
My new course is about distributing Elixir, my way :)

I'm using Swarm for registration and pubsub. I might even do some fault-tolerant stuff if I have time.

I'm bumping into a minor roadblock.

I have an agent-based server:

~~~ elixir
defmodule Dictionary.WordList do

  @me {:via, Swarm, __MODULE__ }

  use Agent

  def start_link(_) do
    Agent.start_link(&word_list/0, name: @me)
  end

  def random_word() do
    Agent.get(@me, &Enum.random/1)
  end

  ...
~~~

When another node connects to the one running this server, Swarm sends a `{:swarm, :begin_handoff}` to the Agent, which crashes it.

Is there a way that I can continue to use an Agent but still register its PID via Swarm?

Cheers

Dave

**Comment by slashdotdash** [comment]
@pragdave I think what you have should work, so it looks like a bug. Let me investigate further.

**Comment by slashdotdash** [comment]
@pragdave Could you try using [Swarm v3.0.5](https://hex.pm/packages/swarm/3.0.5) (`{:swarm, "~> 3.0"}`) and let me know if that works. I believe that version should be ok.

**Comment by slashdotdash** [comment]
@bitwalker I've created a pull request to fix this issue (#63). Shall we merge in and push a 3.1.1 bug fix release to Hex?

**Comment by slashdotdash** [comment]
@bitwalker Might the changes I made in #46 have caused the strange behaviour Dave is encountering?

**Comment by bitwalker** [comment]
@slashdotdash I don't believe #46 caused this issue, at least as far as I can tell. I think Dave is likely correct here, and that the initial synchronization between the nodes when they are connected has a bug. Processes registered with `register_name/2` have no `mfa` metadata correct? I've noticed you've made a number of changes where the function heads now match on that metadata structure, and I'm wondering if perhaps one such change was made somewhere in the sync process where it shouldn't be (or we need to add another clause to handle the case with no `mfa` metadata). I'd have to dig through the commits, but you probably know better than I off the top of your head if any of the code in those areas have been modified recently like that. In any case, I think revisiting the initial synchronization flow will reveal the bug.

**Comment by bitwalker** [comment]
@slashdotdash I would consider the initial sync problem still present even after merging 63, we'll want to address that before pushing a new release.

**Comment by pragdave** [comment]
Let me know if I can help.

**Comment by slashdotdash** [comment]
@bitwalker I've identified the steps to reproduce this issue as follows.

1. Start node1 and register a PID using `Swarm.register_name/2`:

```
13:19:50.966 [info]  [swarm on node1@127.0.0.1] [tracker:init] started
13:19:50.971 [info]  [swarm on node1@127.0.0.1] [tracker:cluster_wait] joining cluster..
13:19:50.971 [info]  [swarm on node1@127.0.0.1] [tracker:cluster_wait] no connected nodes, proceeding without sync
iex(node1@127.0.0.1)1> {:ok, pid} = Agent.start_link(fn -> [] end)
{:ok, #PID<0.222.0>}
iex(node1@127.0.0.1)2> Swarm.register_name("agent", pid)
13:20:01.995 [debug] [swarm on node1@127.0.0.1] [tracker:handle_call] registering #PID<0.222.0> as "agent", with metadata %{}
:yes
iex(node1@127.0.0.1)3> Swarm.registered()
[{"agent", #PID<0.222.0>}]
```

2. Start node2 and wait until after `sync_nodes_timeout` (default 5 seconds) before connecting to node1:

```
13:24:37.438 [info]  [swarm on node2@127.0.0.1] [tracker:init] started
13:24:37.444 [info]  [swarm on node2@127.0.0.1] [tracker:cluster_wait] joining cluster..
13:24:37.444 [info]  [swarm on node2@127.0.0.1] [tracker:cluster_wait] no connected nodes, proceeding without sync
iex(node2@127.0.0.1)2> Node.connect(:"node1@127.0.0.1")
true
13:24:42.950 [info]  [swarm on node2@127.0.0.1] [tracker:ensure_swarm_started_on_remote_node] nodeup node1@127.0.0.1 
13:24:42.951 [debug] [swarm on node2@127.0.0.1] [tracker:handle_topology_change] topology change (nodeup for node1@127.0.0.1)
13:24:42.951 [info]  [swarm on node2@127.0.0.1] [tracker:handle_topology_change] topology change complete
```

3. `Swarm.registered/0` on node2 returns an empty list, it doesn't include the process already registered on node1:

```
iex(node2@127.0.0.1)3> Swarm.registered()              
[]
```

4. Wait until `anti_entropy_interval` has elapsed (default 5 minutes) when the registries are sync'd between the two nodes:

```
13:24:46.035 [debug] [swarm on node2@127.0.0.1] [tracker:handle_cast] received sync request from node1@127.0.0.1
13:24:46.036 [info]  [swarm on node2@127.0.0.1] [tracker:awaiting_sync_ack] received sync acknowledgement from node1@127.0.0.1
13:24:46.036 [info]  [swarm on node2@127.0.0.1] [tracker:resolve_pending_sync_requests] pending sync requests cleared
13:24:47.450 [info]  [swarm on node2@127.0.0.1] [tracker:anti_entropy] syncing with node1@127.0.0.1
13:24:47.450 [info]  [swarm on node2@127.0.0.1] [tracker:syncing] received registry from node1@127.0.0.1, merging..
13:24:47.454 [debug] [swarm on node2@127.0.0.1] [tracker:syncing] local is missing "agent", adding..
13:24:47.454 [info]  [swarm on node2@127.0.0.1] [tracker:syncing] synchronization complete!
13:24:47.454 [info]  [swarm on node2@127.0.0.1] [tracker:resolve_pending_sync_requests] pending sync requests cleared
```

5. Now `Swarm.registered/0` on both nodes returns the registered process:

```
 iex(node2@127.0.0.1)4> Swarm.registered()
[{"agent", #PID<13579.222.0>}]
```

In summary, if a node doesn't connect to another node before the `sync_nodes_timeout` then it won't have a valid registry until after the first anti-entropy interval. The default interval settings means there's a five minute period where the second node is unaware of already registered processes *unless* it tries to register a process using an existing name.

Can you recommend a fix for this? Would it be sensible to use a very short anti-entropy interval when the tracker starts without any other nodes. We could then gradually increase the interval until the configured setting - using an exponential backoff for example. 

An alternative solution is have the tracker remain in the `cluster_wait` state until it actually forms a cluster. At which point it will immediately request a registry from the connected node. I tried this approach and it appeared to work, albeit it's a more significant change to the `Swarm.Tracker` module.

**Comment by pragdave** [comment]
Would it be possible to start the tracker until the nose has at least one
other node connected to it?

**Comment by pragdave** [comment]
This code is way above my pay grade without a serious investment in time, but...

where does `defp handle_topology_change({type, remote_node}, %TrackerState{} = state)` in the tracker actually initiate a sync? It just seems to return a `{:keep_state, lclock}` in this circumstance.

**Comment by bitwalker** [comment]
@pragdave It used to be that a nodeup event would involve a synchronization round automatically, but it looks like that changed at some point (I can't recall if I did so, or if that came through in a PR at some point, but it doesn't do so now).

It looks like (and I haven't confirmed this yet) we just need to ensure that if a node was in the tracking state, and not connected to any other nodes, then we need to handle nodeup events a little differently, basically the same way the tracker comes up if a cluster is formed right at startup, e.g.:

```
# cluster forms during startup
init -> cluster_wait -> cluster_join -> syncing -> tracking

# cluster forms while in tracking state (current behavior)
tracking -> handle_topology_change -> tracking

# cluster forms while in tracking state (desired behavior)
tracking -> cluster_join -> syncing -> tracking
```

The `cluster_join` state implicitly handles the topology change, the only big difference is that it enters the syncing state after updating the state of the tracker with the new node. Without this step, the node joining the cluster doesn't get an opportunity to receive the current state of the registry across the cluster, and it only contains registrations it made itself, or ones received after it was created, at least until the next anti-entropy pass occurs.

**Comment by bitwalker** [comment]
I pushed a PR which should fix this, but needs to be tested. If either of you can run a repro using that branch and confirm that it does fix the issue, I'll go ahead and get it merged.

**Comment by bitwalker** [comment]
As noted in #64 - it fixes the issue, however it uncovers a deficiency in how synchronization is handled, in that multiple tiebreaker die rolls occur at the same time, and we accept whichever one completes first, which results in "unhandled cast" messages for the orphaned die rolls. We should choose a node deterministically to coordinate the tiebreaking process, so that only one is ever needed.

**Comment by pragdave** [comment]
Will you be creating a release with #63 and #64?

**Comment by bitwalker** [comment]
I can do that right away, unless you'd like to wait until I address the unhandled cast issue.

**Comment by pragdave** [comment]
Actually, if you could, it would be tremendous. I'm just about to start recording the coding of this part.

You are a hero. Thank you for all you do.


Dave

**Comment by bitwalker** [comment]
You bet! I've pushed it to GitHub/Hex as 3.2.0 :)

**Comment by pragdave** [comment]
Hugs

**Comment by pragdave** [comment]
Sorry about this, but that warning about the die roll turns out to be hiding a bug.

![image](https://user-images.githubusercontent.com/10648/35660140-f1319834-06cf-11e8-94cf-2a8aeae17462.png)


Here's my guess: there's a race condition during the dice roll. Both sides both win and lose.

~~~
21:38:39.055 [info]  [swarm on a@plasma] [tracker:cluster_wait] selected sync node: b@plasma                        │21:38:39.065 [info]  [swarm on b@plasma] [tracker:cluster_wait] selected sync node: a@plasma
                                                                                                                    │
21:38:39.073 [info]  [swarm on a@plasma] [tracker:syncing] there is a tie between syncing nodes, breaking with die r│21:38:39.074 [info]  [swarm on b@plasma] [tracker:syncing] there is a tie between syncing nodes, breaking wit
oll (19)..                                                                                                          │h die roll (20)..
                                                                                                                    │
21:38:39.074 [info]  [swarm on a@plasma] [tracker:syncing] there is a tie between syncing nodes, breaking with die r│21:38:39.074 [info]  [swarm on b@plasma] [tracker:syncing] there is a tie between syncing nodes, breaking wit
oll (10)..                                                                                                          │h die roll (7)..
                                                                                                                    │
21:38:39.074 [info]  [swarm on a@plasma] [tracker:syncing] b@plasma won the die roll (20 vs 10),                    │21:38:39.074 [info]  [swarm on b@plasma] [tracker:syncing] a@plasma won the die roll (19 vs 7),
                                                                                                                    │
21:38:39.074 [info]  [swarm on a@plasma] [tracker:syncing] we won the die roll (19 vs 7), sending payload..         │21:38:39.075 [info]  [swarm on b@plasma] [tracker:syncing] we won the die roll (20 vs 10), sending payload..
~~~

So both sides send the payload to each other. Quite why this causes the right-hand window's app not to receive the Dictionary.Wordlist, I don't know.

This is with Hex 3.2.0

**Comment by bitwalker** [comment]
Yeah I noticed that when I was investigating, I'll get that addressed tomorrow, sorry for the trouble!

**Comment by bitwalker** [comment]
@pragdave I've pushed a fix to master which I believe addresses a number of issues and edge cases. Could you give it a try and let me know if you run into any problems? I did some testing locally and it appeared to work as expected

@slashdotdash Could you review commit 412bad9 and let me know if you have any concerns? If you have any questions I can explain the changes in more detail

**Comment by pragdave** [comment]
It looks good! Many thanks. I can't believe how responsive you guys have
been.

**Comment by bitwalker** [comment]
No problem! Since it looks good, I will go ahead and get that released as 3.2.1 then :)

I'll close this issue for now, but feel free to either reopen it, or open a new one, if there are still issues with synchronization
